File: apps/runtime/src/terminal/direct.rs
//! The fallback backend: one `portable-pty` per session, held by this process.
//!
//! Identical in behaviour to the pre-§15 `PtyManager` — the session dies with
//! the runtime, `capture` replays the output buffer instead of reading a real
//! screen, and `paste` writes the bracketed-paste sequence straight into the
//! PTY. It is what runs when tmux is missing, too old, or turned off.

use std::{
    collections::{HashMap, VecDeque},
    sync::{
        Arc, Mutex,
        atomic::{AtomicBool, Ordering},
    },
    time::Duration,
};

use async_trait::async_trait;
use bytes::Bytes;
use portable_pty::{Child, CommandBuilder, MasterPty, PtySize, native_pty_system};
use tokio::sync::{RwLock, broadcast, mpsc};

use crate::{
    error::{AppError, AppResult},
    terminal::backend::{
        AttachHandle, BackendKind, BackendNotice, BackendRef, DetachGuard, ForegroundInfo,
        NoticeSender, OutputPump, PASTE_END, PASTE_START, REPLAY_CHUNKS, SessionKey,
        TerminalBackend, TerminalHandle, TerminalSpec, child_commands, child_environment,
        process_table, sanitize_paste, spawn_input_pump, spawn_output_pump, strip_escapes,
        tail_lines, terminate_tree, trim_captured,
    },
};

struct DirectSession {
    generation: u64,
    /// Dropped on exit so the session stops holding a file descriptor while it
    /// keeps serving `capture` for the scrollback the UI still shows.
    master: Mutex<Option<Box<dyn MasterPty + Send>>>,
    child: Mutex<Option<Box<dyn Child + Send + Sync>>>,
    pid: Option<i64>,
    input: mpsc::Sender<Bytes>,
    output: broadcast::Sender<Bytes>,
    replay: Arc<Mutex<VecDeque<Bytes>>>,
    exited: AtomicBool,
}

impl DirectSession {
    fn snapshot(&self) -> String {
        let chunks = self
            .replay
            .lock()
            .map(|replay| replay.iter().cloned().collect::<Vec<_>>())
            .unwrap_or_default();
        let mut joined = Vec::new();
        for chunk in chunks {
            joined.extend_from_slice(&chunk);
        }
        String::from_utf8_lossy(&joined).into_owned()
    }
}

#[derive(Clone)]
pub struct DirectBackend {
    sessions: Arc<RwLock<HashMap<SessionKey, Arc<DirectSession>>>>,
    notices: NoticeSender,
}

impl DirectBackend {
    pub fn new(notices: NoticeSender) -> Self {
        Self {
            sessions: Arc::new(RwLock::new(HashMap::new())),
            notices,
        }
    }

    async fn session(&self, key: &SessionKey) -> AppResult<Arc<DirectSession>> {
        self.sessions
            .read()
            .await
            .get(key)
            .cloned()
            .ok_or_else(|| AppError::NotFound("Terminal session is not running".into()))
    }

    async fn live(&self, key: &SessionKey) -> AppResult<Arc<DirectSession>> {
        let session = self.session(key).await?;
        if session.exited.load(Ordering::SeqCst) {
            return Err(AppError::Conflict("Terminal session has exited".into()));
        }
        Ok(session)
    }

    pub async fn pid(&self, key: &SessionKey) -> Option<i64> {
        let session = self.sessions.read().await.get(key).cloned()?;
        if session.exited.load(Ordering::SeqCst) {
            return None;
        }
        session.pid
    }

    /// The replay buffer as one string — the `snapshot` frame of plan §15.5.
    pub async fn snapshot(&self, key: &SessionKey) -> Option<String> {
        let session = self.sessions.read().await.get(key).cloned()?;
        let snapshot = session.snapshot();
        (!snapshot.is_empty()).then_some(snapshot)
    }
    /// Explicit desktop Quit: report failures and observe real child exit before
    /// removing owned sessions. Normal detach/restart keeps its existing behavior.
    pub async fn shutdown_owned_checked(&self) -> AppResult<()> {
        let sessions: Vec<_> = self
            .sessions
            .read()
            .await
            .iter()
            .map(|(key, session)| (key.clone(), session.clone()))
            .collect();
        let outcomes =
            futures_util::future::join_all(sessions.into_iter().map(|(key, session)| async move {
                let already_exited = {
                    let mut child = session
                        .child
                        .lock()
                        .map_err(|_| AppError::Internal("terminal child lock poisoned".into()))?;
                    match child.as_mut() {
                        Some(child) => child
                            .try_wait()
                            .map_err(|error| AppError::Internal(error.to_string()))?
                            .is_some(),
                        None => true,
                    }
                };
                if !already_exited {
                    if let Some(pid) = session.pid {
                        #[cfg(unix)]
                        terminate_tree(pid).await;
                        #[cfg(windows)]
                        {
                            let result = tokio::process::Command::new("taskkill")
                                .args(["/PID", &pid.to_string(), "/T", "/F"])
                                .kill_on_drop(true)
                                .output()
                                .await?;
                            if !result.status.success() {
                                return Err(AppError::Internal(format!(
                                    "taskkill failed for owned terminal {key}"
                                )));
                            }
                        }
                    }
                    {
                        let mut child = session.child.lock().map_err(|_| {
                            AppError::Internal("terminal child lock poisoned".into())
                        })?;
                        if let Some(child) = child.as_mut()
                            && child
                                .try_wait()
                                .map_err(|error| AppError::Internal(error.to_string()))?
                                .is_none()
                        {
                            child.kill().map_err(|error| {
                                AppError::Internal(format!("could not terminate {key}: {error}"))
                            })?;
                        }
                    }
                    tokio::time::timeout(Duration::from_secs(3), async {
                        loop {
                            let exited = {
                                let mut child = session.child.lock().map_err(|_| {
                                    AppError::Internal("terminal child lock poisoned".into())
                                })?;
                                match child.as_mut() {
                                    Some(child) => child
                                        .try_wait()
                                        .map_err(|error| AppError::Internal(error.to_string()))?
                                        .is_some(),
                                    None => true,
                                }
                            };
                            if exited {
                                return Ok::<(), AppError>(());
                            }
                            tokio::time::sleep(Duration::from_millis(25)).await;
                        }
                    })
                    .await
                    .map_err(|_| {
                        AppError::Internal(format!("owned terminal {key} did not exit"))
                    })??;
                }
                session.exited.store(true, Ordering::SeqCst);
                self.sessions.write().await.remove(&key);
                Ok::<(), AppError>(())
            }))
            .await;
        let failures: Vec<String> = outcomes
            .into_iter()
            .filter_map(Result::err)
            .map(|error| error.to_string())
            .collect();
        if failures.is_empty() {
            Ok(())
        } else {
            Err(AppError::Internal(failures.join("; ")))
        }
    }
}

#[async_trait]
impl TerminalBackend for DirectBackend {
    fn kind(&self) -> BackendKind {
        BackendKind::Direct
    }

    async fn create(&self, spec: TerminalSpec) -> AppResult<TerminalHandle> {
        let executable = spec.executable();
        let pair = native_pty_system()
            .openpty(spec.size)
            .map_err(|error| AppError::Internal(format!("Could not create PTY: {error}")))?;
        let mut command = CommandBuilder::new(&executable);
        command.cwd(&spec.cwd);
        for argument in &spec.args {
            command.arg(argument);
        }
        // A built environment, never the runtime's own (see
        // `child_environment`). Set before `spec.env` so a caller can still
        // override any of it.
        command.env_clear();
        for (name, value) in child_environment() {
            command.env(name, value);
        }
        // Hook variables carry addresses only, never credentials: any process of
        // the same user can read another process' environment.
        for (name, value) in &spec.env {
            command.env(name, value);
        }
        let child = pair.slave.spawn_command(command).map_err(|error| {
            AppError::BadRequest(format!("Could not start {executable}: {error}"))
        })?;
        drop(pair.slave);
        let reader = pair
            .master
            .try_clone_reader()
            .map_err(|error| AppError::Internal(format!("Could not read PTY: {error}")))?;
        let writer = pair
            .master
            .take_writer()
            .map_err(|error| AppError::Internal(format!("Could not write PTY: {error}")))?;
        let pid = child.process_id().map(i64::from);
        let (output, _) = broadcast::channel(512);
        let (input, input_receiver) = mpsc::channel::<Bytes>(256);
        let session = Arc::new(DirectSession {
            generation: spec.generation,
            master: Mutex::new(Some(pair.master)),
            child: Mutex::new(Some(child)),
            pid,
            input,
            output: output.clone(),
            replay: Arc::new(Mutex::new(VecDeque::with_capacity(REPLAY_CHUNKS))),
            exited: AtomicBool::new(false),
        });
        self.sessions
            .write()
            .await
            .insert(spec.session_key.clone(), session.clone());

        spawn_input_pump(spec.session_key.to_string(), writer, input_receiver);
        spawn_output_pump(
            OutputPump {
                name: spec.session_key.to_string(),
                output,
                replay: Some(session.replay.clone()),
                notices: self.notices.clone(),
                key: spec.session_key.clone(),
                generation: spec.generation,
            },
            reader,
            || {},
        );
        self.spawn_exit_watcher(spec.session_key.clone(), session);

        Ok(TerminalHandle {
            session_key: spec.session_key,
            generation: spec.generation,
            backend_ref: None,
            pid,
        })
    }

    async fn attach(
        &self,
        key: &SessionKey,
        generation: u64,
        size: PtySize,
    ) -> AppResult<AttachHandle> {
        let session = self.session(key).await?;
        if session.generation != generation {
            return Err(stale(generation, session.generation));
        }
        if !session.exited.load(Ordering::SeqCst) {
            self.resize(key, size).await?;
        }
        Ok(AttachHandle {
            output: session.output.subscribe(),
            input: session.input.clone(),
            generation: session.generation,
            detach: DetachGuard::none(),
        })
    }

    async fn write(&self, key: &SessionKey, bytes: &[u8]) -> AppResult<()> {
        let session = self.live(key).await?;
        session
            .input
            .send(Bytes::copy_from_slice(bytes))
            .await
            .map_err(|_| AppError::NotFound("Terminal session is not running".into()))
    }

    async fn resize(&self, key: &SessionKey, size: PtySize) -> AppResult<()> {
        let session = self.live(key).await?;
        let master = session
            .master
            .lock()
            .map_err(|_| AppError::Internal("Terminal size controller is unavailable".into()))?;
        let Some(master) = master.as_ref() else {
            return Ok(());
        };
        master
            .resize(PtySize {
                rows: size.rows.max(2),
                cols: size.cols.max(2),
                pixel_width: 0,
                pixel_height: 0,
            })
            .map_err(|error| AppError::Internal(format!("Could not resize PTY: {error}")))
    }

    /// No real screen to read: the replay buffer is the best approximation.
    async fn capture(&self, key: &SessionKey, lines: u32, with_escapes: bool) -> AppResult<String> {
        let session = self.session(key).await?;
        let raw = session.snapshot();
        let text = if with_escapes {
            raw
        } else {
            strip_escapes(&raw)
        };
        Ok(tail_lines(&trim_captured(&text.replace('\r', "")), lines))
    }

    async fn paste(&self, key: &SessionKey, text: &str, press_enter: bool) -> AppResult<()> {
        let mut payload = String::with_capacity(text.len() + 16);
        payload.push_str(PASTE_START);
        payload.push_str(&sanitize_paste(text));
        payload.push_str(PASTE_END);
        if press_enter {
            payload.push('\r');
        }
        self.write(key, payload.as_bytes()).await
    }

    async fn foreground(&self, key: &SessionKey) -> AppResult<ForegroundInfo> {
        let session = self.session(key).await?;
        let Some(pid) = session.pid else {
            return Ok(ForegroundInfo::default());
        };
        let table = process_table();
        Ok(ForegroundInfo {
            pid: Some(pid),
            command: table.get(&pid).map(|(_, argv)| argv.clone()),
            children: child_commands(pid),
        })
    }

    async fn interrupt(&self, key: &SessionKey) -> AppResult<()> {
        // 0x03 travels through the line discipline, which turns it into SIGINT
        // for the foreground process group — the same path as a real Ctrl+C.
        self.write(key, b"\x03").await
    }

    async fn terminate_process(&self, key: &SessionKey) -> AppResult<()> {
        let session = self.session(key).await?;
        if let Some(pid) = session.pid {
            terminate_tree(pid).await;
        }
        if let Ok(mut child) = session.child.lock()
            && let Some(child) = child.as_mut()
        {
            let _ = child.kill();
        }
        Ok(())
    }

    async fn destroy(&self, key: &SessionKey) -> AppResult<()> {
        let _ = self.terminate_process(key).await;
        self.sessions.write().await.remove(key);
        Ok(())
    }

    async fn list_alive(&self) -> AppResult<Vec<BackendRef>> {
        Ok(self
            .sessions
            .read()
            .await
            .iter()
            .filter(|(_, session)| !session.exited.load(Ordering::SeqCst))
            .map(|(key, _)| BackendRef {
                name: key.to_string(),
                attached: true,
            })
            .collect())
    }

    /// Direct sessions cannot survive the runtime, so "release" means "kill".
    async fn detach_all(&self) {
        let keys: Vec<SessionKey> = self.sessions.read().await.keys().cloned().collect();
        for key in keys {
            let _ = self.destroy(&key).await;
        }
    }
}

impl DirectBackend {
    fn spawn_exit_watcher(&self, key: SessionKey, session: Arc<DirectSession>) {
        let notices = self.notices.clone();
        tokio::spawn(async move {
            loop {
                let status = session
                    .child
                    .lock()
                    .ok()
                    .and_then(|mut child| child.as_mut().and_then(|child| child.try_wait().ok()))
                    .flatten();
                if let Some(status) = status {
                    session.exited.store(true, Ordering::SeqCst);
                    // Let the reader drain whatever the process wrote just
                    // before it died, then release the descriptors.
                    tokio::time::sleep(Duration::from_millis(100)).await;
                    if let Ok(mut master) = session.master.lock() {
                        *master = None;
                    }
                    if let Ok(mut child) = session.child.lock() {
                        *child = None;
                    }
                    let _ = notices.send(BackendNotice::Exited {
                        session_key: key,
                        generation: session.generation,
                        exit_code: Some(status.exit_code() as i64),
                    });
                    break;
                }
                tokio::time::sleep(Duration::from_millis(250)).await;
            }
        });
    }
}

fn stale(requested: u64, current: u64) -> AppError {
    AppError::Conflict(format!(
        "Terminal generation {requested} is stale; the session is at {current}"
    ))
}

File: apps/desktop/src/core/terminal/direct.ts
import { StringDecoder } from "node:string_decoder";

import {
  type Attachment,
  type BackendCapabilities,
  type BackendKind,
  type BackendNotice,
  type BackendRef,
  DORMANT_FLUSH_INTERVAL_MS,
  type ForegroundInfo,
  OUTPUT_FLUSH_BYTES,
  OUTPUT_FLUSH_INTERVAL_MS,
  PASTE_END,
  PASTE_START,
  REPLAY_CHUNKS,
  type SessionKey,
  type TerminalBackend,
  type TerminalHandle,
  type TerminalSize,
  type TerminalSpec,
  type TerminateMode,
  conflict,
  executable,
  notFound,
  sanitizePaste,
  tailLines,
  trimCaptured,
} from "./backend";
import { asRecord, childEnvironment } from "./environment";
import { childCommands, processTable, terminateTree } from "./process";
import { type Pty, openPty, releasePty } from "./pty";
import { ReplayScreen } from "./replay-screen";

/**
 * The fallback backend: one PTY per session, held by this process.
 *
 * A direct session **dies with the core**, and every behaviour here follows
 * from that one fact:
 *
 *   * `capture` reads a screen this backend keeps itself, fed with every
 *     byte the PTY writes, because there is no terminal to re-read;
 *   * `attach` owes the socket a `snapshot` frame, because nothing redraws;
 *   * `scroll` does nothing: xterm owns that scrollback on the page side;
 *   * `detachAll` **kills**, because "release without ending the session" is
 *     not a thing a process-owned PTY can do.
 *
 * It is what runs when tmux is missing, too old, or turned off — and on
 * Windows when the session host cannot be reached.
 *
 * ## Why the batching is here and not only at the socket
 *
 * `socket.ts` coalesces on its own 16 ms budget, so a second batcher looks
 * redundant. It is not: a session nobody is attached to has no socket at all,
 * and its output still has to be kept for the next attach. Batching here is
 * what makes the replay ring hold ~128 *screenfuls* instead of ~128 8 KiB PTY
 * reads, and it is the only place `setDormant` has anything to slow down.
 */

/** How long `terminate` waits for a killed child before it stops looking. */
const EXIT_POLL_MS = 25;
const EXIT_TIMEOUT_MS = 3_000;

interface DirectListener {
  readonly id: number;
  onData: ((chunk: Buffer) => void) | undefined;
  onExit: ((exitCode: number | undefined) => void) | undefined;
  /** Bytes that arrived before the socket subscribed are not dropped. */
  readonly buffered: Buffer[];
  subscribed: boolean;
}

interface DirectSession {
  readonly key: SessionKey;
  readonly generation: number;
  pty: Pty | undefined;
  readonly pid: number | undefined;
  readonly listeners: Map<number, DirectListener>;
  /** The last {@link REPLAY_CHUNKS} flushed batches, oldest first. */
  readonly replay: Buffer[];
  /** Accumulating batch, flushed on `cadence` or {@link OUTPUT_FLUSH_BYTES}. */
  pending: Buffer[];
  pendingBytes: number;
  timer: NodeJS.Timeout | undefined;
  cadence: number;
  exited: boolean;
  exitCode: number | undefined;
  /**
   * Everything the PTY has written, laid out on a screen of its current size
   * (`replay-screen.ts`) — what `capture` reads. Fed from the first byte, not
   * rebuilt from the replay ring: a CLI that draws its prompt once and then
   * only redraws a spinner pushes the prompt out of the ring within seconds.
   */
  readonly screen: ReplayScreen;
  readonly decoder: StringDecoder;
}

export class DirectBackend implements TerminalBackend {
  readonly kind: BackendKind = "direct";
  private readonly sessions = new Map<SessionKey, DirectSession>();
  private readonly listenerSinks: ((notice: BackendNotice) => void)[] = [];
  private nextListenerId = 1;

  getCapabilities(): BackendCapabilities {
    return {
      kind: "direct",
      persistent: false,
      // Nothing redraws a PTY this process owns, so the socket is owed the
      // replay buffer as one `snapshot` frame.
      redrawsOnAttach: false,
      usable: true,
    };
  }

  notices(listener: (notice: BackendNotice) => void): void {
    this.listenerSinks.push(listener);
  }

  /* --------------------------------- create ------------------------------- */

  async create(spec: TerminalSpec): Promise<TerminalHandle> {
    const env = asRecord(childEnvironment());
    // The caller's variables last, so an explicit `LANG` or `ARMADRA_*` wins
    // over the built block rather than being shadowed by it.
    for (const [name, value] of spec.env) env[name] = value;

    const pty = openPty({
      file: executable(spec),
      args: spec.args,
      cwd: spec.cwd,
      env,
      cols: spec.size.cols,
      rows: spec.size.rows,
    });

    const session: DirectSession = {
      key: spec.sessionKey,
      generation: spec.generation,
      pty,
      pid: pty.pid,
      listeners: new Map(),
      replay: [],
      pending: [],
      pendingBytes: 0,
      timer: undefined,
      screen: new ReplayScreen(spec.size.cols, spec.size.rows),
      decoder: new StringDecoder("utf8"),
      cadence: OUTPUT_FLUSH_INTERVAL_MS,
      exited: false,
      exitCode: undefined,
    };
    this.sessions.set(spec.sessionKey, session);

    pty.onData((data) => {
      this.absorb(session, Buffer.isBuffer(data) ? data : Buffer.from(data));
    });
    pty.onExit((event) => {
      // Whatever the process wrote just before it died is still in the batch.
      this.flush(session);
      session.exited = true;
      session.exitCode = event.exitCode;
      session.pty = undefined;
      for (const listener of session.listeners.values()) {
        listener.onExit?.(event.exitCode);
      }
      this.announce({
        type: "exited",
        key: session.key,
        generation: session.generation,
        exitCode: event.exitCode,
      });
    });

    return {
      sessionKey: spec.sessionKey,
      generation: spec.generation,
      // No external handle: there is no server holding this session, so there
      // is nothing an orphan sweep could ever address it by.
      pid: pty.pid,
    };
  }

  /* --------------------------------- output ------------------------------- */

  private absorb(session: DirectSession, chunk: Buffer): void {
    session.pending.push(chunk);
    session.pendingBytes += chunk.byteLength;
    if (session.pendingBytes >= OUTPUT_FLUSH_BYTES) {
      this.flush(session);
      return;
    }
    if (session.timer !== undefined) return;
    session.timer = setTimeout(() => {
      this.flush(session);
    }, session.cadence);
    session.timer.unref?.();
  }

  private flush(session: DirectSession): void {
    if (session.timer !== undefined) {
      clearTimeout(session.timer);
      session.timer = undefined;
    }
    if (session.pending.length === 0) return;
    const batch =
      session.pending.length === 1
        ? (session.pending[0] as Buffer)
        : Buffer.concat(session.pending);
    session.pending = [];
    session.pendingBytes = 0;
    session.replay.push(batch);
    while (session.replay.length > REPLAY_CHUNKS) session.replay.shift();
    session.screen.write(session.decoder.write(batch));
    for (const listener of session.listeners.values()) {
      if (!listener.subscribed) {
        listener.buffered.push(batch);
        continue;
      }
      listener.onData?.(batch);
    }
  }

  private announce(notice: BackendNotice): void {
    for (const sink of this.listenerSinks) sink(notice);
  }

  /** The replay buffer as one string — the `snapshot` frame of §15.5. */
  snapshot(key: SessionKey): string | undefined {
    const session = this.sessions.get(key);
    if (session === undefined) return undefined;
    const text = Buffer.concat([...session.replay, ...session.pending])
      .toString("utf8")
      .replace(/\u{FFFD}+$/u, "");
    return text === "" ? undefined : text;
  }

  /* --------------------------------- attach ------------------------------- */

  async attach(
    key: SessionKey,
    generation: number,
    size: TerminalSize,
  ): Promise<Attachment> {
    const session = this.require(key);
    if (session.generation !== generation) {
      throw conflict(
        `Terminal generation ${generation} is stale; the session is at ${session.generation}`,
      );
    }
    // A session that already ended still attaches: the socket needs somewhere
    // to deliver the final `status` from, and the snapshot is still worth
    // showing. Resizing a dead PTY is not, so that is skipped.
    if (!session.exited) await this.resize(key, size);

    const id = this.nextListenerId;
    this.nextListenerId += 1;
    const listener: DirectListener = {
      id,
      onData: undefined,
      onExit: undefined,
      buffered: [],
      subscribed: false,
    };
    session.listeners.set(id, listener);

    return {
      attachmentId: id,
      generation: session.generation,
      onData: (sink) => {
        listener.onData = sink;
        if (listener.subscribed) return;
        listener.subscribed = true;
        for (const chunk of listener.buffered.splice(0)) sink(chunk);
      },
      onExit: (sink) => {
        listener.onExit = sink;
        // The process may have ended between `attach` and this subscription.
        if (session.exited) sink(session.exitCode);
      },
    };
  }

  async detach(key: SessionKey, attachmentId: number): Promise<void> {
    this.sessions.get(key)?.listeners.delete(attachmentId);
  }

  /* ---------------------------------- input ------------------------------- */

  async input(key: SessionKey, bytes: Buffer): Promise<void> {
    const session = this.live(key);
    session.pty?.write(bytes.toString("utf8"));
  }

  /**
   * No paste buffer to load: the bracketed-paste sequence goes straight into
   * the PTY, and the Enter that may follow travels in the **same write** so
   * nothing can interleave between the closing bracket and the newline.
   */
  async paste(
    key: SessionKey,
    text: string,
    pressEnter: boolean,
  ): Promise<void> {
    const payload = `${PASTE_START}${sanitizePaste(text)}${PASTE_END}${pressEnter ? "\r" : ""}`;
    await this.input(key, Buffer.from(payload, "utf8"));
  }

  async resize(key: SessionKey, size: TerminalSize): Promise<void> {
    const session = this.live(key);
    try {
      session.screen.resize(size.cols, size.rows);
      session.pty?.resize(
        Math.max(2, Math.trunc(size.cols)),
        Math.max(2, Math.trunc(size.rows)),
      );
    } catch {
      // A PTY that died between the lookup and the call.
    }
  }

  /**
   * No real screen to re-read, so the plain capture reads the one this backend
   * keeps ({@link DirectSession.screen}): a full-screen CLI draws with cursor
   * moves, not lines, and stripping the escapes off the replay leaves either
   * one run-on line or a tail of redraw fragments. The escaped form is the
   * replay itself.
   */
  async capture(
    key: SessionKey,
    lines: number,
    withEscapes: boolean,
  ): Promise<string> {
    // A session that has produced nothing captures as the empty string; one
    // that does not exist is a 404, and the two must not look alike.
    const session = this.require(key);
    // What arrived since the last batch is part of the screen too.
    this.flush(session);
    const text = withEscapes
      ? (this.snapshot(key) ?? "").replaceAll("\r", "")
      : session.screen.text();
    return tailLines(trimCaptured(text), lines);
  }

  /**
   * 0x03 travels through the line discipline, which turns it into SIGINT for
   * the foreground process group — the same path as a real Ctrl+C. Signalling
   * the pid directly would hit the shell instead of whatever it is running.
   */
  async signal(key: SessionKey, _signal: "interrupt"): Promise<void> {
    await this.input(key, Buffer.from([0x03]));
  }

  async getForeground(key: SessionKey): Promise<ForegroundInfo> {
    const session = this.require(key);
    if (session.pid === undefined) return { children: [] };
    const table = processTable();
    const argv = table.get(session.pid)?.argv;
    return {
      pid: session.pid,
      ...(argv === undefined ? {} : { command: argv }),
      children: childCommands(session.pid, table),
    };
  }

  /* -------------------------------- lifecycle ----------------------------- */

  async terminate(key: SessionKey, mode: TerminateMode): Promise<void> {
    if (mode === "interrupt") {
      await this.signal(key, "interrupt");
      return;
    }
    const session = this.require(key);
    if (session.pid !== undefined) await terminateTree(session.pid);
    releasePty(session.pty);
    session.pty = undefined;
    if (mode === "session") this.sessions.delete(key);
  }

  async list(): Promise<BackendRef[]> {
    const refs: BackendRef[] = [];
    for (const [key, session] of this.sessions) {
      if (session.exited) continue;
      refs.push({ name: key, attached: session.listeners.size > 0 });
    }
    return refs;
  }

  /**
   * A direct session has no handle anything outside this process could hold,
   * so there is no such thing as an orphan of this backend and nothing to
   * destroy by reference. Answering rather than throwing keeps the orphan
   * sweep's one loop written once for every backend.
   */
  async destroyByReference(_reference: string): Promise<void> {}

  /** xterm owns this scrollback; there is no second history to move. */
  async scroll(_key: SessionKey, _lines: number): Promise<void> {}

  /**
   * The process and its replay buffer stay exactly as they are; only the
   * delivery cadence changes. Waking is therefore free and can never be
   * mistaken for a create.
   */
  async setDormant(key: SessionKey, dormant: boolean): Promise<void> {
    const session = this.require(key);
    session.cadence = dormant
      ? DORMANT_FLUSH_INTERVAL_MS
      : OUTPUT_FLUSH_INTERVAL_MS;
  }

  /**
   * Core shutdown. Direct sessions cannot survive it, so "release" means
   * "kill" — the one place where this backend's `detachAll` differs from
   * every other one's, and the reason tmux is the primary backend.
   */
  async detachAll(): Promise<void> {
    for (const key of [...this.sessions.keys()]) {
      try {
        await this.terminate(key, "session");
      } catch {
        // Already gone; shutdown must not stall on one dead session.
      }
    }
  }

  /**
   * Explicit desktop Quit, unlike a core restart: every owned session is
   * stopped and the exit is **observed**, not assumed. Returns the keys it
   * could not prove had ended, so the caller can report an incomplete quit
   * instead of a silent one.
   */
  async shutdownOwnedChecked(): Promise<string[]> {
    const failures: string[] = [];
    for (const [key, session] of [...this.sessions]) {
      if (session.exited) {
        this.sessions.delete(key);
        continue;
      }
      try {
        if (session.pid !== undefined) await terminateTree(session.pid);
        releasePty(session.pty, "SIGKILL");
        const deadline = Date.now() + EXIT_TIMEOUT_MS;
        while (!session.exited && Date.now() < deadline) {
          await new Promise((resolve) => setTimeout(resolve, EXIT_POLL_MS));
        }
        if (!session.exited) {
          failures.push(`owned terminal ${key} did not exit`);
          continue;
        }
      } catch (error) {
        failures.push(
          `could not terminate ${key}: ${error instanceof Error ? error.message : String(error)}`,
        );
        continue;
      }
      this.sessions.delete(key);
    }
    return failures;
  }

  /* --------------------------------- helpers ------------------------------- */

  private require(key: SessionKey): DirectSession {
    const session = this.sessions.get(key);
    if (session === undefined) {
      throw notFound("Terminal session is not running");
    }
    return session;
  }

  private live(key: SessionKey): DirectSession {
    const session = this.require(key);
    if (session.exited) throw conflict("Terminal session has exited");
    return session;
  }
}
